### PR TITLE
Proxy manager shutdown [API-1259] [API-1342]

### PR DESCRIPTION
**Pull request by nevzatseferoglu**
When shutdown happens, already assigned proxies are destroyed and each destroy sends a signal to the server.

### CODE REVIEW

File: internal/client/client.go [hunk 1]
```diff
@@ -86,9 +86,26 @@ func (c *Config) Validate() error {
...
+
+const ProxyShutdownHandler ShutdownHandlerType = iota
+
+func executeShutdownHandlers(
+	ctx context.Context,
+	m map[ShutdownHandlerType]func(ctx context.Context) error) error {
+	for _, f := range m {
+		err := f(ctx)
+		if err != nil {
+			return err
```
Comment (utku-caglayan) [inline internal/client/client.go:99]: I think we should accumulate errors, but also continue to the shutdown process. If something goes wrong, we the only action I can think of is logging. So would you find it useful to accumulate errors and continue to iterating?

Comment (utku-caglayan) [reply internal/client/client.go:99]: Also how about adding ShutdownHandlerType to the error for better context? In that case we may need to somehow convert to a string I guess.

File: proxy_manager.go [hunk 2]
```diff
@@ -222,3 +222,32 @@ func (m *proxyManager) getFlakeIDGeneratorConfig(name string) FlakeIDGeneratorCo
...
+	err := ds.Destroy(ctx)
+	if err != nil {
+		return fmt.Errorf("given %s proxy cannot be destroyed, %s", name, err)
+	}
+	return nil
+}
+
+func (m *proxyManager) destroyProxies(ctx context.Context) error {
+	for key, proxy := range m.proxies {
+		err := destroy(ctx, key, proxy)
```
Comment (utku-caglayan) [inline proxy_manager.go:247]: I think we can inline this `destroy` method. It seems like it doesn't add much flexibility but introduces another error level. WDYT?

**Review by utku-caglayan — COMMENTED** [review]
Thanks! Here are few comments that we can discuss

File: client.go [hunk 3]
```diff
@@ -369,4 +369,8 @@ func (c *Client) createComponents(config *Config) {
 	}
 	c.proxyManager = newProxyManager(proxyManagerServiceBundle)
 	c.sqlService = isql.NewService(c.ic.ConnectionManager, c.ic.SerializationService, c.ic.InvocationFactory, c.ic.InvocationService, &c.ic.Logger)
+	// create and assign shutdown handlers to be executed in internal shutdown
+	shutdownHandlers := make(map[client.ShutdownHandlerType]func(ctx context.Context) error)
```
[the remark this answers is not in the document]
Comment (nevzatseferoglu) [reply client.go]: Thanks, fixed [0dce062](https://github.com/hazelcast/hazelcast-go-client/pull/772/commits/0dce06299b7e0f4ec33083c83fbdf45de299988a).

File: proxy.go [hunk 4]
```diff
@@ -433,6 +433,10 @@ func (p *proxy) sendInvocation(ctx context.Context, inv invocation.Invocation) e
 	return p.invocationService.SendRequest(ctx, inv)
 }
 
+func (p proxy) Name() string {
```
[the remark this answers is not in the document]
Comment (nevzatseferoglu) [reply proxy.go]: Thanks, fixed [0dce062](https://github.com/hazelcast/hazelcast-go-client/pull/772/commits/0dce06299b7e0f4ec33083c83fbdf45de299988a).

File: internal/client/client.go (hunk 1 above)
Comment (nevzatseferoglu) [reply internal/client/client.go]: I created a custom error type for handler but, I'm not sure that I accumulate correctly [0dce062](https://github.com/hazelcast/hazelcast-go-client/pull/772/commits/0dce06299b7e0f4ec33083c83fbdf45de299988a).

File: proxy_manager.go (hunk 2 above)
Comment (nevzatseferoglu) [reply proxy_manager.go]: Thanks, fixed [0dce062](https://github.com/hazelcast/hazelcast-go-client/pull/772/commits/0dce06299b7e0f4ec33083c83fbdf45de299988a).

File: proxy_manager.go [hunk 5]
```diff
@@ -222,3 +224,33 @@ func (m *proxyManager) getFlakeIDGeneratorConfig(name string) FlakeIDGeneratorCo
...
+
+type proxyDestroyer interface {
+	Destroy(ctx context.Context) error
+}
+
+func (m *proxyManager) destroyProxies(ctx context.Context) {
+	for key, p := range m.proxies {
+		if p == nil {
+			m.serviceBundle.Logger.Log(logger.WeightError, func() string {
+				return hzerrors.NewShutdownHandlerError(
```
Comment (utku-caglayan) [inline proxy_manager.go:236]: We don't have to create this error, instead we can log it directly. Also we can also ditch the custom error type. Wdyt?

**Review by utku-caglayan — CHANGES_REQUESTED** [review]

File: proxy_manager.go (hunk 5 above)
Comment (nevzatseferoglu) [reply proxy_manager.go]: - You are right. Since we don't need to identify this error other places. There is no need for creating custom error type.
Fixed, [34fd390](https://github.com/hazelcast/hazelcast-go-client/pull/772/commits/34fd3909b56616404810d1d3e9171fb605f3e865).

**Review by utku-caglayan — DISMISSED** [review]

**Review by utku-caglayan — DISMISSED** [review]
LGTM

File: client_internals.go [hunk 6]
```diff
@@ -85,6 +85,10 @@ func NewClientInternal(c *Client) *ClientInternal {
 	}
 }
 
+func (ci *ClientInternal) ProxyManagerProxies() map[string]interface{} {
```
Comment (yuce) [inline client_internals.go:88]: This looks like it is used only in a test. Could you move it to `client_internals_test_utils.go`?

File: internal/client/client.go [hunk 7]
```diff
@@ -88,11 +88,33 @@ func (c *Config) Validate() error {
...
+		panic(fmt.Errorf("handlerID: %d was already added to shutdown handler before", handlerID))
+	}
+	(*shutdownHandler)[handlerID] = handler
+}
+
+func (shutdownHandler *ShutdownHandler) executeShutdownHandlers(ctx context.Context) {
+	for _, f := range *shutdownHandler {
+		f(ctx)
+	}
+}
```
Comment (yuce) [inline internal/client/client.go:112]: Are these necessary? How about adding the following method:
```
func (c *Client) destroyProxies(ctx context.Context) {
  c.proxyManager.destroyProxies(ctx)
}
```
And then:
```
c.ic.AddShutdownHandler(c.destroyProxies)
```

File: proxy_manager.go [hunk 8]
```diff
@@ -231,3 +232,31 @@ func (m *proxyManager) getFlakeIDGeneratorConfig(name string) FlakeIDGeneratorCo
...
+}
+
+func (m *proxyManager) destroyProxies(ctx context.Context) {
+	for key, p := range m.proxies {
+		if p == nil {
+			m.serviceBundle.Logger.Log(logger.WeightError, func() string {
+				return fmt.Sprintf("proxy %s key cannot be destroyed, given proxy argument is nil", key)
+			})
+			continue
+		}
```
Comment (yuce) [inline proxy_manager.go:247]: Is it feasible that `p == nil`? I think we should remove this check.

File: proxy_manager.go [hunk 9]
```diff
@@ -231,3 +232,31 @@ func (m *proxyManager) getFlakeIDGeneratorConfig(name string) FlakeIDGeneratorCo
...
+		ds, ok := p.(proxyDestroyer)
+		if !ok {
+			m.serviceBundle.Logger.Log(logger.WeightError, func() string {
+				return fmt.Sprintf("proxy %s key cannot be destroyed, proxy argument does not implement proxyDestroyer", key)
+			})
+			continue
+		}
```
Comment (yuce) [inline proxy_manager.go:254]: IMO we also don't need this check. We can simply do:
```
ds := p.(proxyDestroyer)
```

File: proxy_manager.go [hunk 10]
```diff
@@ -231,3 +232,31 @@ func (m *proxyManager) getFlakeIDGeneratorConfig(name string) FlakeIDGeneratorCo
...
+		err := ds.Destroy(ctx)
+		if err != nil {
+			m.serviceBundle.Logger.Log(logger.WeightError, func() string {
```
Comment (yuce) [inline proxy_manager.go:257]: How about:
```
m.serviceBundle.Logger.Errorf("proxy %s key cannot be destroyed: %w", key, err)
```

**Review by yuce — CHANGES_REQUESTED** [review]
Some ideas, continuing to review.

File: internal/client/client.go (hunk 7 above)
Comment (nevzatseferoglu) [reply internal/client/client.go:112]: I think that these are the necessary from the perspective of **readability** and **maintenance**. But I can it can be introducing additional destroyProxies to client method set. [df88887](https://github.com/hazelcast/hazelcast-go-client/pull/772/commits/df8888728cd4a5f4a040179d4df2b0cb77c1c8b6).

File: proxy_manager.go (hunk 9 above)
Comment (nevzatseferoglu) [reply proxy_manager.go]: Thank, you are right, it is not needed. [df88887](https://github.com/hazelcast/hazelcast-go-client/pull/772/commits/df8888728cd4a5f4a040179d4df2b0cb77c1c8b6).

File: proxy_manager.go (hunk 10 above)
Comment (nevzatseferoglu) [reply proxy_manager.go]: Thank, it is better for the verbosity of the logger. [df88887](https://github.com/hazelcast/hazelcast-go-client/pull/772/commits/df8888728cd4a5f4a040179d4df2b0cb77c1c8b6).

File: client_internals.go (hunk 6 above)
Comment (nevzatseferoglu) [reply client_internals.go]: Yes, I moved it. [4f3ef73](https://github.com/hazelcast/hazelcast-go-client/pull/772/commits/4f3ef7364ca9128a95eefad10ab2899294a60607).

File: proxy_manager.go (hunk 8 above)
Comment (yuce) [reply proxy_manager.go:247]: Is there a reason to set an entry to `nil`? I think that can only happen by mistake.

File: internal/client/client.go (hunk 7 above)
Comment (yuce) [reply internal/client/client.go:112]: I think the complexity here is making it less readable and maintainable. Do we need to have different types of shutdown handlers?

File: proxy_manager.go (hunk 8 above)
Comment (nevzatseferoglu) [reply proxy_manager.go]: Removed related section, [694aee3](https://github.com/hazelcast/hazelcast-go-client/pull/772/commits/694aee3b72c6dd7465c5d3bde999c2fdc292f28e).

File: internal/client/client.go (hunk 7 above)
Comment (nevzatseferoglu) [reply internal/client/client.go]: Simplied the design, [694aee3](https://github.com/hazelcast/hazelcast-go-client/pull/772/commits/694aee3b72c6dd7465c5d3bde999c2fdc292f28e).

File: client.go [hunk 11]
```diff
@@ -116,6 +116,9 @@ func newClient(config Config) (*Client, error) {
 	}
 	c.addConfigEvents(&config)
 	c.createComponents(&config)
+	c.ic.ShutdownHandlers = make([]func(ctx context.Context), 0)
```
Comment (yuce) [inline client.go:119]: This is not required.

File: internal/client/client.go [hunk 12]
```diff
@@ -105,10 +103,22 @@ type Client struct {
 	PartitionService     *icluster.PartitionService
 	ClusterService       *icluster.Service
 	name                 string
-	shutdownHandlers     []ShutdownHandler
+	ShutdownHandlers     []func(ctx context.Context)
```
Comment (yuce) [inline internal/client/client.go:106]: Why was `ShutdownHandlers` exported?

File: internal/client/client.go [hunk 13]
```diff
@@ -105,10 +103,22 @@ type Client struct {
...
 	state                int32
 }
 
+func (c *Client) AddShutdownHandler(handler func(ctx context.Context)) {
+	// this is supposed to be called during client initialization, so there's no risk of races.
+	c.ShutdownHandlers = append(c.ShutdownHandlers, handler)
+}
+
+func (c *Client) ExecuteShutdownHandlers(ctx context.Context) {
```
Comment (yuce) [inline internal/client/client.go:115]: Is this method required? Do we expect to call it from somewhere other than the internal client?

File: internal/client/client.go [hunk 14]
```diff
@@ -180,13 +185,14 @@ func (c *Client) Shutdown(ctx context.Context) error {
 		ctx = context.Background()
 	}
 	c.EventDispatcher.Publish(lifecycle.NewLifecycleStateChanged(lifecycle.StateShuttingDown))
+	c.ExecuteShutdownHandlers(ctx)
```
Comment (yuce) [inline internal/client/client.go:188]: Shutdown handlers are called twice, here and below. Why is that required?

**Review by yuce — CHANGES_REQUESTED** [review]

File: internal/client/client.go [hunk 15]
```diff
@@ -88,11 +88,9 @@ func (c *Config) Validate() error {
 	return nil
 }
 
-type ShutdownHandler func(ctx context.Context)
```
Comment (yuce) [inline internal/client/client.go:91]: Why did you remove this type?

Comment (nevzatseferoglu) [reply internal/client/client.go:91]: It was the part of the simplifying process, I put it back, It look useful. [a6b3aef](https://github.com/hazelcast/hazelcast-go-client/pull/772/commits/a6b3aef218b08ee2bee4501ee356f35c96a4e97e).

File: internal/client/client.go (hunk 14 above)
Comment (nevzatseferoglu) [reply internal/client/client.go]: Upcoming loop was redundant. Thanks for that catch. Removed it. [a6b3aef](https://github.com/hazelcast/hazelcast-go-client/pull/772/commits/a6b3aef218b08ee2bee4501ee356f35c96a4e97e).

File: internal/client/client.go (hunk 13 above)
Comment (nevzatseferoglu) [reply internal/client/client.go]: I don't think we expect to call it. I inlined it to where it is being called. [17f8ebc](https://github.com/hazelcast/hazelcast-go-client/pull/772/commits/17f8ebc0e27f5a5d71baf61a7eb5b9f5253c25bf).

File: internal/client/client.go (hunk 12 above)
Comment (nevzatseferoglu) [reply internal/client/client.go]: It was the typo. It should not be exported. Fixed, [17f8ebc](https://github.com/hazelcast/hazelcast-go-client/pull/772/commits/17f8ebc0e27f5a5d71baf61a7eb5b9f5253c25bf).

File: Makefile [hunk 16]
```diff
@@ -31,4 +31,4 @@ doc:
 	godoc -http=localhost:$(PORT)
 
 check:
-	bash ./check.sh
+	bash ./check.sh
```
Comment (yuce) [inline Makefile:34]: I guess this change deletes the newline at the end of the line. Could you revert it?

File: internal/client/client.go [hunk 17]
```diff
@@ -88,11 +88,11 @@ func (c *Config) Validate() error {
...
+type shutdownHandler []func(context.Context)
```
Comment (yuce) [inline internal/client/client.go:91]: Why do we need this change?

File: internal/client/client.go [hunk 18]
```diff
@@ -105,10 +105,15 @@ type Client struct {
 	PartitionService     *icluster.PartitionService
 	ClusterService       *icluster.Service
 	name                 string
-	shutdownHandlers     []ShutdownHandler
+	shutdownHandlers     shutdownHandler
 	state                int32
 }
 
+func (c *Client) AddShutdownHandler(handler ...func(ctx context.Context)) {
```
Comment (yuce) [inline internal/client/client.go:112]: Do we need this change?

File: internal/client/client.go [hunk 19]
```diff
@@ -180,15 +180,16 @@ func (c *Client) Shutdown(ctx context.Context) error {
 		ctx = context.Background()
 	}
 	c.EventDispatcher.Publish(lifecycle.NewLifecycleStateChanged(lifecycle.StateShuttingDown))
+	// execute registered shutdown handlers
+	for _, f := range c.shutdownHandlers {
+		f(ctx)
+	}
```
Comment (yuce) [inline internal/client/client.go:186]: Running shutdown handlers here changes the behaviour. Does this change has a reason?

File: client_internals_it_test.go [hunk 20]
```diff
@@ -383,6 +383,36 @@ func TestClientInternal_EncodeData(t *testing.T) {
 	})
 }
 
+func TestProxyManagerShutdown(t *testing.T) {
+	clientTester(t, func(t *testing.T, smart bool) {
+		ctx := context.Background()
+		tc := it.StartNewClusterWithOptions("proxy-manager-graceful-shutdown", 5701, 1)
+		defer tc.Shutdown()
+		config := tc.DefaultConfig()
```
Comment (yuce) [inline client_internals_it_test.go:391]: Could you use `it.DefaultConfigWithNoSSL` instead?

File: client_internals_it_test.go [hunk 21]
```diff
@@ -383,6 +383,36 @@ func TestClientInternal_EncodeData(t *testing.T) {
...
+		config.Cluster.Unisocket = !smart
+		client := it.MustClient(hz.StartNewClientWithConfig(ctx, config))
```
Comment (yuce) [inline client_internals_it_test.go:393]: Could you defer shutdown the client?

File: client_internals_test_utils.go [hunk 22]
```diff
@@ -54,6 +54,10 @@ func (ci *ClientInternal) SerializationService() *serialization.Service {
 	return ci.client.ic.SerializationService
 }
 
+func (ci *ClientInternal) ProxyManagerProxies() map[string]interface{} {
```
Comment (yuce) [inline client_internals_test_utils.go:57]: How about renaming this function to just `Proxies`?

File: client_internals_test_utils.go [hunk 23]
```diff
@@ -54,6 +54,10 @@ func (ci *ClientInternal) SerializationService() *serialization.Service {
...
+	return ci.client.proxyManager.proxies
```
Comment (yuce) [inline client_internals_test_utils.go:58]: `proxyManager.proxies` is protected with a mutex. It should be used to access and copy the proxies before returning it to the user. i would add the `Proxies()` method to `proxyManager` and call it from here.

File: proxy_manager.go [hunk 24]
```diff
@@ -231,3 +231,16 @@ func (m *proxyManager) getFlakeIDGeneratorConfig(name string) FlakeIDGeneratorCo
 func makeProxyName(serviceName string, objectName string) string {
 	return fmt.Sprintf("%s%s", serviceName, objectName)
 }
+
+type proxyDestroyer interface {
+	Destroy(ctx context.Context) error
+}
+
+func (m *proxyManager) destroyProxies(ctx context.Context) {
+	for key, p := range m.proxies {
```
Comment (yuce) [inline proxy_manager.go:240]: `proxies` are protected with a mutex. That should be used in this function before accessing that.

**Review by yuce — CHANGES_REQUESTED** [review]

File: proxy_manager.go (hunk 24 above)
Comment (nevzatseferoglu) [reply proxy_manager.go:251]: Resolved, [0042502](https://github.com/hazelcast/hazelcast-go-client/pull/772/commits/0042502d165af4bbfd57b99474f816dd147c4599).

File: client_internals_test_utils.go (hunk 23 above)
Comment (nevzatseferoglu) [reply client_internals_test_utils.go]: Resolved, [0042502](https://github.com/hazelcast/hazelcast-go-client/pull/772/commits/0042502d165af4bbfd57b99474f816dd147c4599).

File: client_internals_test_utils.go (hunk 22 above)
Comment (nevzatseferoglu) [reply client_internals_test_utils.go]: Resolved, [0042502](https://github.com/hazelcast/hazelcast-go-client/pull/772/commits/0042502d165af4bbfd57b99474f816dd147c4599).

File: client_internals_it_test.go (hunk 21 above)
Comment (nevzatseferoglu) [reply client_internals_it_test.go:394]: Resolved, [0042502](https://github.com/hazelcast/hazelcast-go-client/pull/772/commits/0042502d165af4bbfd57b99474f816dd147c4599).

File: client_internals_it_test.go (hunk 20 above)
Comment (nevzatseferoglu) [reply client_internals_it_test.go]: Resolved, [0042502](https://github.com/hazelcast/hazelcast-go-client/pull/772/commits/0042502d165af4bbfd57b99474f816dd147c4599).

File: internal/client/client.go (hunk 19 above)
Comment (nevzatseferoglu) [reply internal/client/client.go]: Resolved, [0042502](https://github.com/hazelcast/hazelcast-go-client/pull/772/commits/0042502d165af4bbfd57b99474f816dd147c4599).

File: internal/client/client.go (hunk 18 above)
Comment (nevzatseferoglu) [reply internal/client/client.go]: Resolved, [0042502](https://github.com/hazelcast/hazelcast-go-client/pull/772/commits/0042502d165af4bbfd57b99474f816dd147c4599).

File: internal/client/client.go (hunk 17 above)
Comment (nevzatseferoglu) [reply internal/client/client.go]: Resolved, [0042502](https://github.com/hazelcast/hazelcast-go-client/pull/772/commits/0042502d165af4bbfd57b99474f816dd147c4599).

File: Makefile (hunk 16 above)
Comment (nevzatseferoglu) [reply Makefile:34]: Resolved, [0042502](https://github.com/hazelcast/hazelcast-go-client/pull/772/commits/0042502d165af4bbfd57b99474f816dd147c4599).

Comment (nevzatseferoglu) [reply Makefile]: Resolved, [bee4be0](https://github.com/hazelcast/hazelcast-go-client/pull/772/commits/bee4be09cc55d2003a3c84ff7cdbac923c4185f9).

**Review by yuce — APPROVED** [review]
Looks good!

**Review by utku-caglayan — APPROVED** [review]
